### PR TITLE
[SPARK-29795][CORE] Explicitly clear registered metrics on MetricSystem shutdown

### PR DESCRIPTION
**Pull request by srowen**
### What changes were proposed in this pull request?

Explicitly clear registered metrics when `MetricsSystem` shuts down.

### Why are the changes needed?

See https://issues.apache.org/jira/browse/SPARK-29795 for a complete explanation. The TL;DR is there is some evidence this could leak resources after Spark is shut down, and that may be a minor issue in Spark 3+ for apps or tests that re-start SparkContexts in the same JVM.

### Does this PR introduce any user-facing change?

The possible difference here is that, after Spark is stopped, metrics are no longer available. It's unclear to me whether this is intended behavior anyway.

### How was this patch tested?

See https://issues.apache.org/jira/browse/SPARK-29795 for more context:
- Spark 3 already passes tests without this change
- Spark 2.4 does too, as exists in branch-2.4 now
- Spark 2.4 fails tests if metrics 4.x is used, without this change

The last point is not directly relevant, as Spark 2.4 will not use metrics 4.x. It's evidence that it addresses some potential issue, however.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala
Comment (srowen) [inline core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala]: Note that in Spark 2.4, this had to be:
```
registry.removeMatching(new MetricFilter {
  override def matches(name: String, metric: Metric): Boolean = true
})
```

**Review by srowen — COMMENTED** [review]
CC @holdenk and @LucaCanali who may be interested.

**Comment by dongjoon-hyun** [comment]
Thank you, @srowen !

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending Jenkins). Thank you!

**Comment by srowen** [comment]
Let's hold off a bit merging this, just as I want to collect as much feedback as possible that we're not missing something.

BTW I do not think this has to be back-ported to 2.4

**Comment by dongjoon-hyun** [comment]
Got it, @srowen .

**Review by dbtsai — APPROVED** [review]
LGTM.

**Comment by srowen** [comment]
Merged to master

**Comment by dongjoon-hyun** [comment]
Thank you all!

**Pull request closed without merging**
